src/libstore/local-store.cc

```diff
@@ -944,8 +954,11 @@ void LocalStore::registerValidPaths(const ValidPathInfos & infos)
            not be valid yet. */
         for (auto & [_, i] : infos)
             if (i.path.isDerivation()) {
-                // FIXME: inefficient; we already loaded the derivation in addValidPath().
-                readInvalidDerivation(i.path).checkInvariants(*this, i.path);
+                if (auto drv = derivations.find(i.path); drv != derivations.end()) {
+                    drv->second.checkInvariants(*this, i.path);
+                } else {
+                    readInvalidDerivation(i.path).checkInvariants(*this, i.path);
+                }
             }
```

Comment:
Do we even need this code at all? isn't `addValidPath` already calling `checkInvariants`? The code (before you did anything, to be clear) seems redundant to me.

Reply:
It appears to be the other way around. addValidPath is never called with checkOutputs=true so it never checkInvariants.

Reply:
I think you are right. The comment was out-dated.